Show pprof labels in thread names

stefanhaller opened the pull request:
If a goroutine has labels, we add them after the ID.

Fixes #3493.

aarzilli commented on service/dap/server.go:
```diff
@@ -1800,10 +1800,21 @@ func (s *Session) onThreadsRequest(request *dap.ThreadsRequest) {
+				}
+				sort.Strings(keys)
+				for _, k := range keys {
+					labels += fmt.Sprintf(" %s=%s", k, g.Labels()[k])
```
use a `strings.Builder`

stefanhaller replied:
Ah yes, thanks. See 93d379b059a2.

aarzilli reviewed (commented):
cc @hyangah @suzmue

stefanhaller commented:
After testing this for a while I found that `k1:v1` looks better than `k1=v1`, so I pushed a fixup (515f4cdc09bc) to change that. Not a big thing, let me know if you disagree.

derekparker commented:
@hyangah or @suzmue would you mind taking a look at this?

stefanhaller commented:
@derekparker Please note that over in #3493 some concerns were brought up; showing all labels might be too much when there are many. Please see the discussion there, starting at https://github.com/go-delve/delve/issues/3493#issuecomment-1734352570.

suzmue commented:
I think this looks great but would love to see this behind an opt-in configuration to start! (commented on #3493 as well)

alexandear commented on service/dap/server.go:
```diff
@@ -1800,10 +1800,21 @@ func (s *Session) onThreadsRequest(request *dap.ThreadsRequest) {
 			if g.Thread != nil && g.Thread.ThreadID() != 0 {
 				thread = fmt.Sprintf(" (Thread %d)", g.Thread.ThreadID())
 			}
+			var labels strings.Builder
+			if len(g.Labels()) > 0 {
+				var keys []string
+				for k := range g.Labels() {
+					keys = append(keys, k)
+				}
+				sort.Strings(keys)
+				for _, k := range keys {
+					fmt.Fprintf(&labels, " %s:%s", k, g.Labels()[k])
+				}
+			}
```
Can we rewrite it as follows:
```suggestion
			keys := make([]string, 0, len(g.Labels()))
			for k := range g.Labels() {
				keys = append(keys, k)
			}
			sort.Strings(keys)
			var labels strings.Builder
			for _, k := range keys {
				labels.WriteByte(' ')
				labels.WriteString(k)
				labels.WriteByte(':')
				labels.WriteString(g.Labels()[k])
			}
```
?

stefanhaller replied:
The implementation has changed after the discussion in #3493, as we don't show all labels any more.

I'm curious though, why is writing individual bytes manually better than Fprintf to a builder?

stefanhaller commented:
@suzmue I force-pushed to implement the config, please have a look!

suzmue commented:
I think having a more general solution here could be better. I think having the flexibility to look at all the pprof labels would be great.

We could combine that with another config to only show specific labels but I don't think that needs to go into this change, but I either a list of labels to show or a string that can be parsed like goroutineFilters would be good. It would be nice to be able to show both a single label like the name use case, and a combination of labels that are interesting to the user.

We can also make the pprof labels config easier to adjust / more discoverable by adding it to the callstack context menu in VS Code. Once we get this change in here I am happy to add it (see [hide system goroutines example](https://github.com/golang/vscode-go/issues/1797#issuecomment-1762145244))

stefanhaller commented:
@suzmue Now I'm confused. 😅 Could you please be more specific as to what you would like me to do in this PR?

It sounds like you are thinking of having two different configs, where the second could be added later and would be some sort of filter for the labels. But then what's the first one? A bool option to enable the feature? But once we add the filter, would we still need the bool option?

I'm happy to implement the second option now, whatever it is, but I'd like to have agreement over the exact design before I start, in order to avoid wasting time. My proposal would be to add a single config, `showPprofLabels`, which is a list of strings, default empty. No globbing or regex, just a simple list of keys, because that simplifies things a lot. As a special case for my use case, I'd show just the value of the label if the length of the array is one, and `k:v` otherwise. The downside of this approach is that there's no easy way to say "show all keys" without listing them all manually; but we could add the special case of a single element of `"*"` for that, if needed.

Thoughts?

suzmue commented:
@stefanhaller I think your idea of showPprofLabels as a list of strings sounds great! Those two special cases also sound good to me. Thanks!

stefanhaller commented:
@suzmue I force pushed to implement the proposed solution.

stefanhaller commented:
Ping. @suzmue, would you have time to have a look at this? (Or anybody else?)

suzmue commented on service/dap/config.go:
```diff
@@ -85,3 +94,37 @@ func configureSetSubstitutePath(args *launchAttachArgs, rest string) error {
 	return nil
 }
+
+func configureSetShowPprofLabels(args *launchAttachArgs, rest string) error {
```
Could you add documentation to command.go explaining how this configuration works please? For msgConfig

stefanhaller replied:
Added in 285116f8256e

stefanhaller commented on service/dap/server.go:
```diff
@@ -1812,10 +1822,41 @@ func (s *Session) onThreadsRequest(request *dap.ThreadsRequest) {
 				thread = fmt.Sprintf(" (Thread %d)", g.Thread.ThreadID())
 			}
+			var labels strings.Builder
+			appendLabelsForKeys := func(keys []string) {
```
Renamed in 49a08989663e

stefanhaller commented on service/dap/server_test.go:
```diff
@@ -1386,10 +1393,94 @@ func TestSelectedThreadsRequest(t *testing.T) {
+						if len(tr.Body.Threads) != 1 {
+							t.Errorf("got %d threads, expected 1\n", len(tr.Body.Threads))
+						}
+						// The first breakpoint is before the call to pprof.Do; no lables yet:
```
Fixed in fb5bcaf9234e

stefanhaller commented on service/dap/server_test.go:
```diff
@@ -1386,10 +1393,94 @@ func TestSelectedThreadsRequest(t *testing.T) {
+		showPprofLabelsConfig   []string
+		expectedPrefixWithLabel string
+	}{
+		{
```
Sure, reformatted in d5ac1c646505

stefanhaller commented:
Thanks for the review, I addressed all requests. I referenced the individual fixup commits above to make it easier to see what I changed, but I'm going to squash them now (and rebase on master again).
